includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,356 @@
     $archival_file = islandora_video_copy_obj_datastream($object);
     if ($archival_file === FALSE) {
       return islandora_video_no_obj_datastream($object->id);
     }
     $filePath = drupal_realpath($archival_file['file']->uri);
     $failureMessage = array(
       'success' => false,
-      'messages' => array(
-        'message' => t('Failed to upload object @pid to Panopto, but no exception caught.'),
+      'messages' => array(),
+    );
+    $streaming = array();
+    
+    $oauthendpoint = variable_get('islandora_video_streaming_panopto_host');
+    $uploadendpoint = $oauthendpoint . '/Panopto/PublicAPI/Rest/sessionUpload';
+    $oauthendpoint .= "/Panopto/oauth2/connect/token";
+    $token = islandora_video_streaming_panopto_get_token($oauthendpoint);
+    $session = null;
+    if ($token) {
+      $session = islandora_video_streaming_panopto_get_session($token, $uploadendpoint); 
+    }
+    else {
+      $messages = array(
+        'message' => t('Failed to retrieve token for object: @pid'),
         'message_sub' => array('@pid' => $object->id),
         'type' => 'watchdog',
         'severity' => WATCHDOG_ERROR,
-      ),
-    );
-
-    //TODO: upload, record result in $streaming
- 
+      );
+      $failureMessage['messages'][] = $messages;
+      return $failureMessage;
+    }
+    $success = false;
+    if ($session) {
+      try {
+        $upload_id = $session['ID'];
+        $upload_target = $session['UploadTarget'];
+        $file_uri = explode("/", $filePath);
+        $FILENAME = end($file_uri);
+        
+        islandora_video_streaming_panopto_create_manifest_for_video($filePath, MANIFEST_FILE_NAME);
+        islandora_video_streaming_panopto_upload_file($upload_target, MANIFEST_FILE_NAME);
+        islandora_video_streaming_panopto_upload_file($upload_target, $filePath);
+        islandora_video_streaming_panopto_finish_upload($uploadendpoint, $session, $token);
+        
+        try {
+          $success = monitor_progress($uploadendpoint, $upload_id, $token);
+        } catch (Exception $ex) {
+          $messages = array(
+          'message' => t('Panopto progress error: @exception'),
+          'message_sub' => array('@exception' => $ex->getMessage()),
+          'type' => 'watchdog',
+          'severity' => WATCHDOG_ERROR,
+          );
+          $failureMessage['messages'][] = $messages;
+          watchdog_exception('islandora_video_streaming_panopto', $ex);
+          return $failureMessage;
+        }
+        $streaming['upload_id'] = $session['ID'];
+      } catch (Exception $ex) {
+        $messages = array(
+          'message' => t('Panopto upload error: @exception'),
+          'message_sub' => array('@exception' => $ex->getMessage()),
+          'type' => 'watchdog',
+          'severity' => WATCHDOG_ERROR,
+        );
+        $failureMessage['messages'][] = $messages;
+        watchdog_exception('islandora_video_streaming_panopto', $ex);
```

Comment:
Elsewhere you return early from the function on error, but not here.  What is the difference?  Note still that any early return bypasses your file cleanup (mentioned above).

Reply:
I thought if the function does not return the failuremessage then it would continue onto `file_delete($archival_file['file'], true);`,

Reply:
Yes it will.  I was warning against just adding a `return` here as the solution.